docs/edsl.md

```diff
@@ -0,0 +1,735 @@
+
+First, index validity is determined for a full set of index variables: `j = 1`
+is not valid or invalid as a standalone index value, but may be part of a valid
+or invalid set of index variables. For example, in the code:
+
+```c++
+I.bind_dims(N);
+auto O = TensorOutput((N + 1) / 2);
+O(i) >= I(2 * i + j);
+O.add_constraints({j<2});
+}
+```
+
+with `N = 5`, the indices `i = 1, j = 1` are valid indices.
+However, `i = 2, j = 1` are not valid indices for this operation, nor are `i = -1000, j = 1`.
+
+A set of indices are _valid_ if and only if:
+
+1. All the index variables are integers.
+1. All the index expressions for every tensor are in range. Specifically, if the
+   index variable values are plugged into every index expression, all the
+   resulting indices are non-negative integers less than the appropriate
+   dimension.
+1. All the constraints are satisfied.
+   Constraints always take the form `[index expression] < [constant expression]`
+   (where `[index expression]` is a linear polynomial in the index
+   variables and `[constant expression]` is a linear polynomial in the input
+   dimensions), and they always implicitly include `0 <= [index expression]`.
+   Therefore we could also state this requirement as "every constraint's index
+   expression is non-negative and less than its specified upper bound".
+
+### Skipping
+
+The rule that all index variables must be integers allows us to "skip" certain
+otherwise valid entries. For example, consider the Tile function:
+
+```c++
+Tensor skip(const Tensor& I) {
+  TensorDim M, N;
+  TensorIndex i, j;
+  I.bind_dims(M, N);
+  auto O = TensorOutput(N);
+  O(2 * i) += I(2 * i, j);
+  return O;
+}
+```
+
+This operation only writes to even entries of `O`; while `i = 1/2, j = 1` does
+yield valid index expressions (`O[1]` and `I[1, 1]`), using a fractional index
+variable `i` makes these indices invalid. Note that some elements of `O` are
+never written to. Any unwritten elements in the output of a contraction are
+initialized to `0`.
+
+### Cumulative Sum
+
+Suppose we want to take the cumulative sum of a 1D tensor. That is, we want
+`O[i]` to be the sum of all input entries `I[k]` where `k <= i`. In summation
+notation, this is:
+
+```math
+\Large
+O[i] = \sum_{k \leq i} I[k]
+```
+
+However, we can't use `k <= i` as a constraint in Tile; all the index variables
+must be gathered into a single index expression on one side of the inequality.
+Thus, we rewrite this as `0 <= i - k`. Since the `0` bound is implicitly included
+in all constraints, we just need to choose an upper bound large enough to never
+be hit. From the dimensions of the tensors, we already know `i < N` and `0 <= k`,
+and so `N` is an appropriate upper bound. The resulting Tile code is:
+
+```c++
+Tensor csum(const Tensor& I) {
+  TensorDim N;
+  TensorIndex i, k;
+  I.bind_dims(N);
+  auto O = TensorOutput(N);
+  O(i) += I(k);
+  O.add_constraints ({i - k < N})
```

Comment:
This is missing an ending semicolon and also there's a space between in front of the open paren.

Reply:
Good catch. I can modify these. Thanks :)